apps/explorer/src/hooks/useSearch.ts

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+    isValidTransactionDigest,
+    type SuiTransactionResponse,
+    isValidSuiAddress,
+    isValidSuiObjectId,
+    normalizeSuiObjectId,
+    is,
+    SuiObject,
+} from '@mysten/sui.js';
+import { useQuery } from '@tanstack/react-query';
+
+import { useRpc } from '~/hooks/useRpc';
+import { isGenesisLibAddress } from '~/utils/api/searchUtil';
+
+export function useSearch(query: string) {
+    const rpc = useRpc();
+
+    return useQuery(
+        ['search', query],
+        () => {
+            const promises = [];
+
+            if (isValidTransactionDigest(query)) {
+                promises.push(async () => {
+                    const txdata: SuiTransactionResponse =
+                        await rpc.getTransactionWithEffects(query);
+                    return [
+                        {
+                            label: 'transaction',
+                            results: [
+                                {
+                                    id: txdata.certificate.transactionDigest,
+                                    label: txdata.certificate.transactionDigest,
+                                    type: 'transaction',
+                                },
+                            ],
+                        },
+                    ];
+                });
+            }
+
+            if (isValidSuiAddress(query) && !isGenesisLibAddress(query)) {
+                promises.push(async () => {
+                    const [from, to] = await Promise.all([
+                        rpc.getTransactions({ FromAddress: query }, null, 1),
+                        rpc.getTransactions({ ToAddress: query }, null, 1),
+                    ]);
+
+                    if (from.data[0] || to.data[0]) {
+                        return [
+                            {
+                                label: 'address',
+                                results: [
+                                    {
+                                        id: query,
+                                        label: query,
+                                        type: 'address',
+                                    },
+                                ],
+                            },
+                        ];
+                    } else {
+                        throw new Error('not a valid address');
+                    }
+                });
+            }
+
+            const normalized = normalizeSuiObjectId(query);
+            if (isValidSuiObjectId(normalized)) {
+                promises.push(async () => {
```

Comment:
I feel it _might_ approach here is to break these out into functions like `getResultsForObject` which can just return `null` if the query isn't valid. Then this function can just return something like:

```ts
return Promise.all([getResultsForTransaction(), getResultsForObject()]).filter(Boolean)
```

Just trying to think through the easiest way to maintain this, and I'm not convinced this promise list approach is the best.

Reply:
yea - i was between breaking this out and not... I feel like it's right on the line. also not convinced about the approach